src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/evolutionaryAlgorithm/EvolutionaryAlgorithmQueryOptimizer.java

```diff
@@ -15,90 +15,96 @@
 
         return annotatePlansWithCost(currentGen);
     }
 
-    protected List<PhysicalPlanWithCost> generateTheNextGen( final List<PhysicalPlanWithCost> currentGen ) throws CostEstimationException {
-        List<PhysicalPlan> newCandidates = new ArrayList<>();
+    protected List<PhysicalPlanWithCost> generateNextGen( final List<PhysicalPlanWithCost> currentGen, final RuleApplicationsOfPlan cache ) throws QueryOptimizationException {
+        final Set<PhysicalPlanWithCost> parentSet = new HashSet<>( currentGen );
+        final List<PhysicalPlan> newCandidates = new ArrayList<>();
 
-        while ( newCandidates.size() < nmCandidates ) {
+        while ( newCandidates.size() < nmCandidates && parentSet.size() > 0 ) {
             // pick a parent from the current generation
-            final RandomizedSelection<PhysicalPlanWithCost> planRandomizedSelect = new RandomizedSelection<>();
-            final PhysicalPlan parent = ( planRandomizedSelect.pickOne( new HashSet<>(currentGen) )).getPlan();
+            final PhysicalPlan parent = ( (PhysicalPlanWithCost)randomizedSelection.pickOne( parentSet ) ).getPlan() ;
 
             // determine all possible applications of rewriting rules for the parent plan
-            Set<RuleApplication> ruleApps = ruleApplicationCache.determineRuleApplications( parent );
+            final Set<RuleApplication> ruleApps = cache.determineRuleApplications( parent );
 
-            final RandomizedSelection<RuleApplication> ruleRandomizedSelect = new RandomizedSelection<>();
             if ( ruleApps.size() > 0) {
                 // Pick one rewriting rule based on priority
-                final RuleApplication ruleApplication = ruleRandomizedSelect.pickOne( ruleApps );
+                final RuleApplication ruleApplication = (RuleApplication) randomizedSelection.pickOne( ruleApps );
                 newCandidates.add( ruleApplication.getResultingPlan() );
-                ruleApplicationCache.removeRuleApplications( parent, ruleApplication );
+                ruleApps.remove(ruleApplication);
             }
+            parentSet.remove( parent );
         }
         currentGen.addAll( annotatePlansWithCost(newCandidates) );
 
         // select the next generation from all candidates
         return selectNextGenFromCandidates( currentGen );
     }
 
-    protected List<PhysicalPlanWithCost> annotatePlansWithCost( final List<PhysicalPlan> plans ) throws CostEstimationException {
-        final Double[] costs = CostEstimationUtils.getEstimates( costModel, plans );
+    protected List<PhysicalPlanWithCost> annotatePlansWithCost( final List<PhysicalPlan> plans ) throws QueryOptimizationException {
+        final Double[] costs;
+        try {
+            costs = CostEstimationUtils.getEstimates( ctxt.getCostModel(), plans );
+        } catch ( CostEstimationException e) {
+            throw new QueryOptimizationException("Unexpected exception when determining the cost for plans.", e);
```

Comment:
I would not say that the exception would be totally unexpected here. In contrast, we may have to expect that something goes wrong during the cost estimation. Hence, eventually we need to make the code here more robust and be able to actually deal the case of such an exception rather than just throwing an exception itself. However, this is something for a future PR ;-)

For the moment, simply change the message of the new exception being thrown to the following...

Reply:
```suggestion
            throw new QueryOptimizationException("Determining the cost for plans caused an exception.", e);
```